metagraph/src/annotation/annotation_converters.cpp

```diff
@@ -1032,41 +1038,58 @@ std::unique_ptr<RowDiffAnnotator> convert_to_row_diff(const graph::DBGSuccinct &
                 if (diffs.size() - start_idx >= rows[i].size()) {
                     diffs.resize(start_idx);
                     diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
-                    terminal[anno_ids[i]] = 1;
+                    set_bit(terminal.data(), path[i], true);
                     forced_terminal_count.fetch_add(1, std::memory_order_relaxed);
                 }
                 pos[anno_ids[i]] = { start_idx, diffs.size() - start_idx };
             }
-            pos[anno_ids.back()] = { diffs.size(), rows.back().size() };
-            diffs.insert(diffs.end(), rows.back().begin(), rows.back().end());
+
+            if (!anchor) {  // add the last node as a terminal node
+                terminal_count.fetch_add(1, std::memory_order_relaxed);
+                pos[anno_ids.back()] = { diffs.size(), rows.back().size() };
+                diffs.insert(diffs.end(), rows.back().begin(), rows.back().end());
+            }
+
             size_t offset;
             #pragma omp critical
             {
                 offset = node_diffs.size();
                 node_diffs.insert(node_diffs.end(), diffs.begin(), diffs.end());
             }
-            for (uint32_t i = 0; i < rows.size(); ++i) {
+            for (uint32_t i = 0; i < path.size(); ++i) {
                 pos[anno_ids[i]].first += offset;
             }
 
-            terminal_count.fetch_add(1, std::memory_order_relaxed);
-            terminal[anno_ids.back()] = 1;
-            // add the last row plus one termination bit for each row
-            boundary_size.fetch_add(diffs.size() + rows.size(), std::memory_order_relaxed);
-        },
-        num_threads, false, nullptr, true);
+            // for each node in path, the boundary bit vector has one 0 for each diff and
+            // ends with a 1
+            boundary_size.fetch_add(diffs.size() + path.size(), std::memory_order_relaxed);
+        }, num_threads, max_depth, &terminal);
 
     logger->trace(
             "Traversal done. \n\tTotal rows: {}\n\tTotal diff length: {}\n\t"
             "Avg diff length: {}\n\tTerminal nodes total/max-depth/forced {}/{}/{}\n\t"
             "Avg path length: {}",
-            terminal.size(), node_diffs.size(), 1.0 * node_diffs.size() / terminal.size(),
+            nnodes, node_diffs.size(), 1.0 * node_diffs.size() / terminal.size(),
             terminal_count + forced_terminal_count + depth_terminal_count,
             depth_terminal_count, forced_terminal_count, (double)nnodes / terminal_count);
 
     logger->trace(" Building succinct data structures...");
-    // add the number of nodes that were not visited (the dummy nodes)
+
```

Comment:
after the parallel section is done, there should be a thread fence here to prevent atomic writes from being rearranged after non-atomic reads
```suggestion

std::atomic_thread_fence(std::memory_order_acquire);
```

Reply:
Thanks! Maybe Misha's right and it's better to just use the ++ operator :)